File: admin-panel/lib/api/reservations.ts
import { api } from "../api";
import { getCookie } from "../auth";

export interface SlotInfo {
  id: string;
  slot_datetime: string;
  slot_time_local: string;
}

export interface ChannelInfo {
  id: string;
  name: string;
}

export interface CountryInfo {
  name: string;
  flag_emoji: string | null;
}

export interface UserInfo {
  id: string;
  public_user_code: string;
  full_name: string;
  gender: string | null;
  country: CountryInfo | null;
  participation_score: number;
}

export type AttendanceStatus = "attended" | "absent";

export interface ReservationItem {
  id: string;
  status: string;
  notes: string | null;
  slot: SlotInfo;
  channel: ChannelInfo;
  user: UserInfo;

  // The retired no-show penalty only. Always meant -1 and cannot be
  // re-decided, so it is rendered as history, never as an outcome an admin
  // chose.
  no_show_applied: boolean;

  // The attendance decision. All five are null until an admin records one —
  // which is why `attendance_score_delta` is `number | null` and not `number`:
  // 0 is a real decision ("attended, worth nothing") and must not look like
  // "not judged yet".
  attendance_status: AttendanceStatus | null;
  attendance_score_delta: number | null;
  attendance_reason: string | null;
  attendance_marked_by: string | null;
  attendance_marked_at: string | null;
}

export type ReservationDetail = ReservationItem;

export interface TimelineEvent {
  type: string;
  title: string;
  timestamp: string;
  metadata: Record<string, unknown>;
}

export interface RecordAttendanceBody {
  attendance_status: AttendanceStatus;
  score_delta: number;
  reason: string;
}

export interface AttendanceResponse {
  reservation_id: string;
  user_id: string;
  attendance_status: AttendanceStatus;
  score_delta: number;
  reason: string;
  new_score: number;
  transaction_id: string;
  marked_by: string;
  marked_at: string;
}

// Mirrors ATTENDANCE_SCORE_LIMIT / ATTENDANCE_REASON_MAX in
// app/services/reservation.py. Duplicated because the panel is a separate
// build with no access to them; the server re-checks both and is the
// authority, so the worst a drift can do is a 422 the modal already renders.
export const ATTENDANCE_SCORE_LIMIT = 1000;
export const ATTENDANCE_REASON_MAX = 256;

export interface DaySummary {
  total: number;
  active: number;
  completed: number;
  cancelled: number;
  // "Recorded as absent" under either system — the retired penalty flag or an
  // `absent` decision. Kept under the old name on the wire.
  no_show: number;
  attended: number;
  // Completed, unjudged, not already scored by the retired penalty.
  awaiting_decision: number;
}

export type AttendanceFilter = "" | "pending" | "decided" | "attended" | "absent";

export interface PaginatedReservations {
  items: ReservationItem[];
  total: number;
  page: number;
  pages: number;
  summary: DaySummary;
}

// Mirrors ChannelOut from /admin/channels. `capacity` used to be listed here
// but the endpoint has never returned it — the Channel.capacity column is
// deprecated and a channel's real daily capacity is the number of slots
// generated for it that day.
export interface ChannelItem {
  id: string;
  name: string;
  telegram_channel_id: number;
  priority: number;
  is_active: boolean;
}

export interface ReservationsParams {
  date?: string;
  date_from?: string;
  date_to?: string;
  channel_id?: string;
  status?: string;
  attendance?: string;
  search?: string;
  page?: number;
  page_size?: number;
}

export interface ExportParams {
  date_from: string;
  date_to: string;
  channel_id?: string;
  status?: string;
  attendance?: string;
  format?: "csv" | "json";
}

export async function getReservations(
  params: ReservationsParams = {}
): Promise<PaginatedReservations> {
  const q = new URLSearchParams();
  if (params.date) q.set("date", params.date);
  if (params.date_from) q.set("date_from", params.date_from);
  if (params.date_to) q.set("date_to", params.date_to);
  if (params.channel_id) q.set("channel_id", params.channel_id);
  if (params.status) q.set("status", params.status);
  if (params.attendance) q.set("attendance", params.attendance);
  if (params.search) q.set("search", params.search);
  if (params.page) q.set("page", String(params.page));
  if (params.page_size) q.set("page_size", String(params.page_size));
  const { data } = await api.get<PaginatedReservations>(`/admin/reservations?${q}`);
  return data;
}

export async function getReservation(id: string): Promise<ReservationDetail> {
  const { data } = await api.get<ReservationDetail>(`/admin/reservations/${id}`);
  return data;
}

export async function getReservationTimeline(
  id: string
): Promise<TimelineEvent[]> {
  const { data } = await api.get<TimelineEvent[]>(
    `/admin/reservations/${id}/timeline`
  );
  return data;
}

export async function cancelReservation(
  id: string,
  reason?: string
): Promise<ReservationDetail> {
  const { data } = await api.post<ReservationDetail>(
    `/admin/reservations/${id}/cancel`,
    { reason: reason?.trim() ? reason.trim() : null }
  );
  return data;
}

/** Record an attendance decision and the score it carries.
 *
 * Replaces `markNoShow`, which posted to the now-deprecated `/no-show`
 * endpoint and always meant -1. That function is gone rather than kept
 * alongside: the two mechanisms score the same session and the server refuses
 * whichever comes second, so a panel that could still reach the old one would
 * only ever produce a 409 an admin cannot act on.
 *
 * One decision per reservation, ever. Corrections go through an admin score
 * adjustment on the user.
 */
export async function recordAttendance(
  id: string,
  body: RecordAttendanceBody
): Promise<AttendanceResponse> {
  const { data } = await api.post<AttendanceResponse>(
    `/admin/reservations/${id}/attendance`,
    body
  );
  return data;
}

export async function getChannels(): Promise<ChannelItem[]> {
  const { data } = await api.get<ChannelItem[]>("/admin/channels");
  return data;
}

export async function getAvailableSlots(
  channelId: string,
  date: string
): Promise<SlotInfo[]> {
  const q = new URLSearchParams({ channel_id: channelId, date });
  const { data } = await api.get<SlotInfo[]>(
    `/admin/reservations/available-slots?${q}`
  );
  return data;
}

export async function createReservation(body: {
  user_id: string;
  slot_id: string;
}): Promise<ReservationDetail> {
  const { data } = await api.post<ReservationDetail>("/admin/reservations", body);
  return data;
}

export async function exportReservations(params: ExportParams): Promise<void> {
  const q = new URLSearchParams();
  q.set("date_from", params.date_from);
  q.set("date_to", params.date_to);
  if (params.channel_id) q.set("channel_id", params.channel_id);
  if (params.status) q.set("status", params.status);
  if (params.attendance) q.set("attendance", params.attendance);
  q.set("format", params.format ?? "csv");

  const token = getCookie("admin_access_token");
  const response = await fetch(`/api/admin/reservations/export?${q}`, {
    headers: token ? { Authorization: `Bearer ${token}` } : {},
  });

  if (!response.ok) {
    const text = await response.text();
    throw new Error(text || "Export failed");
  }

  const blob = await response.blob();
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  const ext = params.format === "json" ? "json" : "csv";
  a.download = `reservations_${params.date_from}_${params.date_to}.${ext}`;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(url);
}

File: app/api/admin/reservations.py
import csv
import io
import json
import uuid
from datetime import date, datetime
from math import ceil
from typing import Optional

import pytz
from fastapi import APIRouter, Depends, HTTPException, Query, status
from fastapi.responses import StreamingResponse
from sqlalchemy.ext.asyncio import AsyncSession
from starlette.requests import Request

from app.api.admin.deps import get_current_admin
from app.api.admin.schemas.reservations import (
    AttendanceResponse,
    CancelReservationBody,
    CreateReservationBody,
    DaySummary,
    NoShowResponse,
    PaginatedReservations,
    RecordAttendanceBody,
    ReservationDetail,
    ReservationItem,
    SlotInfo,
    ChannelInfo,
    UserInfo,
    CountryInfo,
)
from app.cache.client import redis_client
from app.core.config import settings
from app.core.exceptions import (
    AttendanceAlreadyRecordedError,
    AttendanceNotDecidableError,
    DailyLimitError,
    DuplicateSlotTimeError,
    LegacyNoShowRecordedError,
    MaxReservationsError,
    NotFoundError,
    PastSlotError,
    ReservationNotCancellableError,
    SameDayCutoffError,
    SlotUnavailableError,
    UserBannedError,
    ValidationError,
)
from app.core.logging import get_logger
from app.db.models.reservation import Reservation, ReservationStatus
from app.db.session import get_db_session
from app.repositories.admin_audit_log import AdminAuditLogRepository
from app.repositories.reservation import (
    ATTENDANCE_FILTERS,
    ReservationRepository,
    _parse_notes,
)
from app.repositories.slot import SlotRepository
from app.schedulers.jobs.score_notification import enqueue_score_notification
from app.services.reservation import ReservationService
from app.services.reservation_timeline import ReservationTimelineService, TimelineEvent
from app.services.score import ParticipationScoreService

router = APIRouter(tags=["admin-reservations"])
logger = get_logger(__name__)

TZ = pytz.timezone(settings.TIMEZONE)

_VALID_STATUS = {s.value for s in ReservationStatus}
_VALID_ATTENDANCE = set(ATTENDANCE_FILTERS)


def _not_found() -> HTTPException:
    return HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Reservation not found")


def _check_attendance_filter(value: Optional[str]) -> None:
    """Reject an unknown attendance filter instead of ignoring it.

    Silently dropping it would return an unfiltered page that looks like a
    filtered one — the worst outcome for a queue view an admin works through.
    """
    if value and value not in _VALID_ATTENDANCE:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=f"Invalid attendance filter. Valid: {sorted(_VALID_ATTENDANCE)}",
        )


def _to_item(res: Reservation) -> ReservationItem:
    slot = res.slot
    slot_local = slot.slot_datetime.astimezone(TZ)
    user = res.user
    country = (
        CountryInfo(
            name=user.country_rel.name,
            flag_emoji=user.country_rel.flag_emoji,
        )
        if user.country_rel
        else None
    )
    return ReservationItem(
        id=res.id,
        status=res.status,
        notes=res.notes,
        slot=SlotInfo(
            id=slot.id,
            slot_datetime=slot.slot_datetime,
            slot_time_local=slot_local.strftime("%H:%M"),
        ),
        channel=ChannelInfo(id=res.channel.id, name=res.channel.name),
        user=UserInfo(
            id=user.id,
            public_user_code=user.public_user_code,
            full_name=user.full_name,
            gender=user.gender,
            country=country,
            participation_score=user.participation_score,
        ),
        no_show_applied=_parse_notes(res.notes).get("no_show_penalty_applied") is True,
        attendance_status=res.attendance_status,
        attendance_score_delta=res.attendance_score_delta,
        attendance_reason=res.attendance_reason,
        attendance_marked_by=res.attendance_marked_by,
        attendance_marked_at=res.attendance_marked_at,
    )


def _today_local() -> date:
    return datetime.now(TZ).date()


# ── GET /export must be declared BEFORE /{reservation_id} ─────────────────────

@router.get("/export")
async def export_reservations(
    date_from: date = Query(...),
    date_to: date = Query(...),
    channel_id: Optional[uuid.UUID] = Query(None),
    status_filter: Optional[str] = Query(None, alias="status"),
    attendance: Optional[str] = Query(None),
    format: str = Query("csv"),
    session: AsyncSession = Depends(get_db_session),
    _admin: str = Depends(get_current_admin),
) -> StreamingResponse:
    if (date_to - date_from).days > 90:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Export range must not exceed 90 days",
        )
    if status_filter and status_filter not in _VALID_STATUS:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=f"Invalid status. Valid: {sorted(_VALID_STATUS)}",
        )
    _check_attendance_filter(attendance)

    repo = ReservationRepository(session)
    rows = await repo.admin_export(
        date_from=date_from,
        date_to=date_to,
        channel_id=channel_id,
        status=status_filter,
        attendance=attendance,
    )

    items = [_to_item(r) for r in rows]

    if format == "json":
        import json as _json
        content = _json.dumps([i.model_dump(mode="json") for i in items], indent=2)
        return StreamingResponse(
            iter([content]),
            media_type="application/json",
            headers={
                "Content-Disposition": (
                    f'attachment; filename="reservations_{date_from}_{date_to}.json"'
                )
            },
        )

    # CSV
    output = io.StringIO()
    writer = csv.writer(output)
    writer.writerow([
        "reservation_id",
        "slot_datetime",
        "slot_time_local",
        "channel_name",
        "user_public_code",
        "user_full_name",
        "country",
        "gender",
        "score_at_export",
        "status",
        # Appended, not inserted: existing consumers read this file by column
        # position as often as by header.
        "no_show_applied",
        "attendance_status",
        "attendance_score",
        "attendance_reason",
        "attendance_marked_by",
        "attendance_marked_at",
    ])
    for item in items:
        country_str = ""
        if item.user.country:
            flag = item.user.country.flag_emoji or ""
            country_str = f"{flag} {item.user.country.name}".strip()
        writer.writerow([
            str(item.id),
            item.slot.slot_datetime.astimezone(TZ).isoformat(),
            item.slot.slot_time_local,
            item.channel.name,
            item.user.public_user_code,
            item.user.full_name,
            country_str,
            item.user.gender or "",
            item.user.participation_score,
            item.status,
            str(item.no_show_applied).lower(),
            item.attendance_status or "",
            # Written out explicitly rather than with `or ""`, which would turn
            # a real decision of 0 points into a blank cell — the one value
            # this feature exists to make expressible.
            "" if item.attendance_score_delta is None else item.attendance_score_delta,
            item.attendance_reason or "",
            item.attendance_marked_by or "",
            item.attendance_marked_at.astimezone(TZ).isoformat()
            if item.attendance_marked_at
            else "",
        ])

    output.seek(0)
    return StreamingResponse(
        iter([output.getvalue()]),
        media_type="text/csv",
        headers={
            "Content-Disposition": (
                f'attachment; filename="reservations_{date_from}_{date_to}.csv"'
            )
        },
    )


# ── GET /available-slots must be declared BEFORE /{reservation_id} ────────────

@router.get("/available-slots", response_model=list[SlotInfo])
async def list_available_slots(
    channel_id: uuid.UUID = Query(...),
    date_param: date = Query(..., alias="date"),
    session: AsyncSession = Depends(get_db_session),
    _admin: str = Depends(get_current_admin),
) -> list[SlotInfo]:
    """Open (un-booked, future) slots for a channel on a date — for the admin
    Create-Reservation slot picker.

    Unlike the user-facing flow this is NOT subject to the channel-unlock
    capacity threshold: an admin picks an explicit channel and should see every
    open slot on it. The same-day cutoff is still enforced at booking time by
    the shared booking core, so a same-day slot shown here may still be rejected
    on submit (surfaced as a clear error)."""
    repo = SlotRepository(session)
    now = datetime.now(TZ)
    slots = await repo.get_available_slots_for_date_and_channel(
        date_param, channel_id, now
    )
    return [
        SlotInfo(
            id=slot.id,
            slot_datetime=slot.slot_datetime,
            slot_time_local=slot.slot_datetime.astimezone(TZ).strftime("%H:%M"),
        )
        for slot in slots
    ]


@router.get("", response_model=PaginatedReservations)
async def list_reservations(
    date_param: Optional[date] = Query(None, alias="date"),
    date_from: Optional[date] = Query(None),
    date_to: Optional[date] = Query(None),
    channel_id: Optional[uuid.UUID] = Query(None),
    status_filter: Optional[str] = Query(None, alias="status"),
    attendance: Optional[str] = Query(None),
    search: Optional[str] = Query(None),
    page: int = Query(1, ge=1),
    page_size: int = Query(100, ge=1, le=500),
    session: AsyncSession = Depends(get_db_session),
    _admin: str = Depends(get_current_admin),
) -> PaginatedReservations:
    """One page of reservations, plus the counts the filter bar shows.

    `attendance` narrows the page to a stage of the attendance decision:
    `pending` (the queue — completed and unjudged), `decided`, `attended` or
    `absent`. The summary counts ignore it on purpose; they are the totals the
    filters are pressed against.
    """
    if status_filter and status_filter not in _VALID_STATUS:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=f"Invalid status. Valid: {sorted(_VALID_STATUS)}",
        )
    _check_attendance_filter(attendance)

    # Default to today when no date args supplied
    date_single = date_param
    if date_single is None and date_from is None and date_to is None:
        date_single = _today_local()

    repo = ReservationRepository(session)

    rows, total = await repo.admin_list(
        date_single=date_single,
        date_from=date_from,
        date_to=date_to,
        channel_id=channel_id,
        status=status_filter,
        search=search,
        attendance=attendance,
        page=page,
        page_size=page_size,
    )

    summary_dict = await repo.admin_summary(
        date_single=date_single,
        date_from=date_from,
        date_to=date_to,
        channel_id=channel_id,
        search=search,
    )

    return PaginatedReservations(
        items=[_to_item(r) for r in rows],
        total=total,
        page=page,
        pages=max(1, ceil(total / page_size)),
        summary=DaySummary(**summary_dict),
    )


@router.post("", response_model=ReservationDetail, status_code=status.HTTP_201_CREATED)
async def create_reservation(
    body: CreateReservationBody,
    request: Request,
    session: AsyncSession = Depends(get_db_session),
    admin: str = Depends(get_current_admin),
) -> ReservationDetail:
    """Admin-create a reservation for an existing user.

    Goes through the same booking core as user booking — identical validation
    (daily limit, max active, duplicate time, double-booking, past-slot,
    same-day cutoff). Booking does not change the score; that happens when an
    admin records attendance afterwards. Banned users are rejected. A
    confirmation DM is sent to the user out-of-band after this request commits.
    """
    svc = ReservationService(session, redis_client)
    try:
        reservation = await svc.admin_create_reservation(
            user_id=body.user_id, slot_id=body.slot_id, actor=admin
        )
    except NotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=exc.message)
    except UserBannedError as exc:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=exc.message)
    except (
        DailyLimitError,
        DuplicateSlotTimeError,
        MaxReservationsError,
        SlotUnavailableError,
    ) as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=exc.message)
    except (PastSlotError, SameDayCutoffError) as exc:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=exc.message)

    audit_repo = AdminAuditLogRepository(session)
    ip = request.client.host if request.client else None
    await audit_repo.log(
        action="reservation_created_by_admin",
        admin_username=admin,
        entity_type="reservation",
        entity_id=str(reservation.id),
        details=f"user_id={body.user_id} slot_id={body.slot_id}",
        ip_address=ip,
    )
    logger.info(
        "admin_reservation_created",
        reservation_id=str(reservation.id),
        admin=admin,
    )

    # Re-load with country eager-loaded so the detail response serializes cleanly.
    detail = await ReservationRepository(session).get_reservation_admin_detail(reservation.id)
    return ReservationDetail(**_to_item(detail or reservation).model_dump())


@router.get("/{reservation_id}", response_model=ReservationDetail)
async def get_reservation(
    reservation_id: uuid.UUID,
    session: AsyncSession = Depends(get_db_session),
    _admin: str = Depends(get_current_admin),
) -> ReservationDetail:
    repo = ReservationRepository(session)
    res = await repo.get_reservation_admin_detail(reservation_id)
    if not res:
        raise _not_found()
    return ReservationDetail(**_to_item(res).model_dump())


@router.get("/{reservation_id}/timeline", response_model=list[TimelineEvent])
async def get_reservation_timeline(
    reservation_id: uuid.UUID,
    session: AsyncSession = Depends(get_db_session),
    _admin: str = Depends(get_current_admin),
) -> list[TimelineEvent]:
    """Read-only chronological timeline for a reservation, oldest first.

    Built on the fly from existing data (reservation, score ledger,
    notification logs, admin audit logs) — no timeline persistence. Returns an
    empty array when the reservation has no derivable events beyond creation.
    """
    svc = ReservationTimelineService(session)
    try:
        return await svc.build(reservation_id)
    except NotFoundError:
        raise _not_found()


@router.post("/{reservation_id}/cancel", response_model=ReservationDetail)
async def cancel_reservation(
    reservation_id: uuid.UUID,
    body: CancelReservationBody,
    request: Request,
    session: AsyncSession = Depends(get_db_session),
    admin: str = Depends(get_current_admin),
) -> ReservationDetail:
    """Admin-cancel an ACTIVE reservation, audit the action, and notify the user.

    No score penalty is applied (operational action). The user DM is sent
    out-of-band after this request commits and cannot roll the cancellation back.
    """
    svc = ReservationService(session, redis_client)
    try:
        res = await svc.admin_cancel_reservation(
            reservation_id, actor=admin, reason=body.reason
        )
    except NotFoundError:
        raise _not_found()
    except ReservationNotCancellableError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=exc.message)

    audit_repo = AdminAuditLogRepository(session)
    ip = request.client.host if request.client else None
    await audit_repo.log(
        action="reservation_cancelled_by_admin",
        admin_username=admin,
        entity_type="reservation",
        entity_id=str(reservation_id),
        details=f"reason={body.reason!r}",
        ip_address=ip,
    )
    logger.info(
        "admin_reservation_cancelled",
        reservation_id=str(reservation_id),
        admin=admin,
    )

    # Re-load with country eager-loaded so the detail response serializes cleanly.
    detail = await ReservationRepository(session).get_reservation_admin_detail(reservation_id)
    return ReservationDetail(**_to_item(detail or res).model_dump())


@router.post("/{reservation_id}/attendance", response_model=AttendanceResponse)
async def record_attendance(
    reservation_id: uuid.UUID,
    body: RecordAttendanceBody,
    request: Request,
    session: AsyncSession = Depends(get_db_session),
    admin: str = Depends(get_current_admin),
) -> AttendanceResponse:
    """Record whether a completed reservation was attended, and what it scored.

    Replaces `POST /{id}/no-show`. The outcome and the score are independent:
    the admin decides both, and no combination is rejected — an attended
    session may be worth nothing and an absence may still be worth points. The
    reason is required because it is quoted back to the user in the DM that
    follows.

    At most one decision per reservation, enforced by a conditional UPDATE
    rather than a read-then-write, and applied before the score so a failure
    can never leave a charge without a decision.

    409 covers the three ways a reservation can be undecidable, and they are
    worth telling apart in the message: it is not COMPLETED yet (a session
    whose slot has just passed stays ACTIVE until the lifecycle job runs at
    :00/:30 — the admin has to wait, not retry differently), a decision already
    exists, or the legacy no-show penalty already scored this session.
    """
    svc = ReservationService(session, redis_client)
    try:
        outcome = await svc.record_attendance(
            reservation_id,
            attendance_status=body.attendance_status,
            score_delta=body.score_delta,
            reason=body.reason,
            actor=admin,
        )
    except NotFoundError:
        raise _not_found()
    except (
        AttendanceAlreadyRecordedError,
        AttendanceNotDecidableError,
        LegacyNoShowRecordedError,
    ) as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=exc.message)
    except ValidationError as exc:
        # The request schema enforces the same bounds, so this is reachable only
        # if the two ever drift — the service is the authority either way.
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=exc.message
        )

    audit_repo = AdminAuditLogRepository(session)
    ip = request.client.host if request.client else None
    await audit_repo.log(
        action="attendance_recorded",
        admin_username=admin,
        entity_type="reservation",
        entity_id=str(reservation_id),
        details=(
            f"status={outcome.attendance_status} "
            f"score_delta={outcome.score_delta:+d} "
            f"reason={outcome.reason!r} tx_id={outcome.transaction_id}"
        ),
        ip_address=ip,
    )
    logger.info(
        "admin_attendance_recorded",
        reservation_id=str(reservation_id),
        attendance_status=outcome.attendance_status,
        score_delta=outcome.score_delta,
        admin=admin,
    )

    return AttendanceResponse(
        reservation_id=outcome.reservation.id,
        user_id=outcome.reservation.user_id,
        attendance_status=outcome.attendance_status,
        score_delta=outcome.score_delta,
        reason=outcome.reason,
        new_score=outcome.new_score,
        transaction_id=outcome.transaction_id,
        marked_by=outcome.reservation.attendance_marked_by,
        marked_at=outcome.reservation.attendance_marked_at,
    )


@router.post(
    "/{reservation_id}/no-show", response_model=NoShowResponse, deprecated=True
)
async def mark_no_show(
    reservation_id: uuid.UUID,
    request: Request,
    session: AsyncSession = Depends(get_db_session),
    admin: str = Depends(get_current_admin),
) -> NoShowResponse:
    """Deprecated — use `POST /{id}/attendance` instead.

    Kept alive only because the shipped admin panel still calls it; the
    replacement records the outcome and the score as separate decisions instead
    of hardcoding -1, and does it atomically.

    Two known defects, which are why it is not being extended: the penalty is
    applied *before* the flag that records it is written, so a crash in between
    charges the user for a decision no later request can see was already made;
    and there is no lock or conditional update, so two concurrent requests both
    read an unflagged row and both charge.
    """
    repo = ReservationRepository(session)
    audit_repo = AdminAuditLogRepository(session)

    res = await repo.get_reservation_admin_detail(reservation_id)
    if not res:
        raise _not_found()

    if res.status != ReservationStatus.COMPLETED:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="No-show can only be applied to completed reservations",
        )

    # The new attendance decision already scored this session. Mirrors the
    # guard record_attendance makes in the other direction, so the two systems
    # cannot both charge one reservation during the changeover.
    if res.attendance_status is not None:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=(
                "An attendance decision has already been recorded for this "
                "reservation; use the attendance endpoint to see it."
            ),
        )

    notes_dict = _parse_notes(res.notes)
    if notes_dict.get("no_show_penalty_applied") is True:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="No-show penalty already applied for this reservation",
        )

    # Apply score penalty
    score_svc = ParticipationScoreService(session)
    tx = await score_svc.apply_no_show_penalty(
        user_id=res.user_id,
        reservation_id=res.id,
        reason="No-show penalty applied by admin",
    )

    # Merge flag into notes JSON
    notes_dict["no_show_penalty_applied"] = True
    res.notes = json.dumps(notes_dict)
    await repo.save(res)

    # Refresh user to get updated score
    await session.refresh(res.user)

    ip = request.client.host if request.client else None
    await audit_repo.log(
        action="no_show_applied",
        admin_username=admin,
        entity_type="reservation",
        entity_id=str(reservation_id),
        details=f"user_id={res.user_id} tx_id={tx.id}",
        ip_address=ip,
    )
    logger.info(
        "admin_no_show_applied",
        reservation_id=str(reservation_id),
        user_id=str(res.user_id),
        admin=admin,
    )

    # Out-of-band DM (delayed job runs after this request's commit lands).
    enqueue_score_notification(tx.id, tx.transaction_type)

    return NoShowResponse(
        reservation_id=res.id,
        user_id=res.user_id,
        new_score=res.user.participation_score,
        transaction_id=tx.id,
    )
